extensions/rust/src/route_map.rs

```diff
@@ -0,0 +1,454 @@
+    pub fn new() -> Self {
+        Self {
+            components: HashSet::new(),
+            children: HashMap::new(),
+            path_parameters: None,
+            asgi_handlers: None,
+            is_asgi: false,
+            static_path: None,
+        }
+    }
+
+    /// Converts the Rust representation to a PyDict representation
+    pub fn as_pydict(&self) -> PyResult<Py<PyDict>> {
+        let gil = Python::acquire_gil();
+        let dict = PyDict::new(gil.python());
+
+        dict.set_item("components", self.components.clone())?;
+        if let Some(ref asgi_handlers) = self.asgi_handlers {
+            dict.set_item("asgi_handlers", asgi_handlers)?;
+        }
+
+        if let Some(ref path_parameters) = self.path_parameters {
+            dict.set_item("path_parameters", path_parameters)?;
+        }
+
+        dict.set_item("is_asgi", self.is_asgi)?;
+
+        if let Some(ref static_path) = self.static_path {
+            dict.set_item("static_path", static_path)?;
+        }
+
+        Ok(dict.into())
+    }
+}
+
+/// A path router based on a prefix tree / trie.
+///
+/// Stores a handler references and other metadata for each node,
+/// as well as both static paths and plain routes for use with Starlite.
+///
+/// Routes can be added using `add_routes`.
+/// Given a scope containing a path, can retrieve handlers using `parse_scope_to_route`.
+#[pyclass]
+pub struct RouteMap {
+    map: Node,
+    static_paths: HashSet<String>,
+    plain_routes: HashSet<String>,
+    starlite: Py<PyAny>,
+    ctx: StarliteContext,
+}
+
+// The functions below are available to Python code
+#[pymethods]
+impl RouteMap {
+    /// Creates an empty `RouteMap`
+    #[new]
+    pub fn new(starlite: Py<PyAny>) -> PyResult<Self> {
+        macro_rules! get_attr_and_downcast {
+            ($module:ident, $attr:expr, $downcast_ty:ty) => {{
+                $module.getattr($attr)?.downcast::<$downcast_ty>()?.into()
+            }};
+        }
+
+        let gil = Python::acquire_gil();
+        let py = gil.python();
+
+        let parsers = py.import("starlite.parsers")?;
+        let parse_path_params = get_attr_and_downcast!(parsers, "parse_path_params", PyFunction);
+
+        let routes = py.import("starlite.routes")?;
+        let http_route = get_attr_and_downcast!(routes, "HTTPRoute", PyType);
+        let web_socket_route = get_attr_and_downcast!(routes, "WebSocketRoute", PyType);
+        let asgi_route = get_attr_and_downcast!(routes, "ASGIRoute", PyType);
+
+        Ok(RouteMap {
+            map: Node::new(),
+            plain_routes: HashSet::new(),
+            static_paths: HashSet::new(),
+            starlite,
+            ctx: StarliteContext {
+                http_route,
+                web_socket_route,
+                asgi_route,
+                parse_path_params,
+            },
+        })
+    }
+
+    /// Adds a new static path by path name
+    pub fn add_static_path(&mut self, path: &str) -> PyResult<()> {
```

Comment:
If the method can't fail, you don't have to return PyResult

Reply:
Fixed this for all of the other methods too